### PR TITLE
feat: Proposed data sync protocols.

### PR DESCRIPTION
**Pull request by richburdon**

### CODE REVIEW

File: packages/core/protocols/src/proto/dxos/data.proto [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+// - Only one Space per sync session.
+// - ISSUE: Can the client can have multiple sessions open at once.
+//      - If not how does the server support multiple clients?
+//      - Can different clients request different sync states?
+//
+// Notes:
+// - https://github.com/dxos/dxos/issues/2227
+
+service DataService {
+  rpc CreateSubscription(DataQuery) returns (stream DataEvent);
```
Comment (dmaretskyi) [inline packages/core/protocols/src/proto/dxos/data.proto:29]: Can a query be updated? E.g. more object ids added to the filter?

File: packages/core/protocols/src/proto/dxos/data.proto [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+message DataQuery {
```
Comment (dmaretskyi) [inline packages/core/protocols/src/proto/dxos/data.proto:32]: how do I control whether I want to get live mutations or not?

File: packages/core/protocols/src/proto/dxos/data.proto [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    NOOP = 0;
+    /// Remove from cache (stop tracking; e.g., deleted or no longer matches filter).
+    REMOVE = 1;
+  }
+
+  string object_id = 1;
+  Action action = 2;
+
+  /// Genesis of object.
+  optional ObjectGenesis genesis = 10;
```
Comment (dmaretskyi) [inline packages/core/protocols/src/proto/dxos/data.proto:69]: Do ServiceWorker -> App events also contain this section to initialize new objects?

File: packages/core/protocols/src/proto/dxos/data.proto [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+message ObjectGenesis {
+  dxos.echo.timeframe.TimeframeVector timestamp = 1;
+  string model_type = 2;
+  optional string model_version = 3;
+  optional string parent_id = 4;
+}
+
+message ObjectSnapshot {
+  dxos.echo.timeframe.TimeframeVector timestamp = 1;
+  google.protobuf.Any payload = 2;
```
Comment (dmaretskyi) [inline packages/core/protocols/src/proto/dxos/data.proto:85]: duplication between this and ObjectGenesis?

File: packages/core/protocols/src/proto/dxos/data.proto (hunk 2 above)
Comment (richburdon) [reply packages/core/protocols/src/proto/dxos/data.proto:32]: i can't come up for a use case of static snapshots, but id as that as a different rpc method.

File: packages/core/protocols/src/proto/dxos/data.proto (hunk 3 above)
Comment (richburdon) [reply packages/core/protocols/src/proto/dxos/data.proto:69]: no. genesis is just on the feed. it's an historical event.

File: packages/core/protocols/src/proto/dxos/data.proto (hunk 4 above)
Comment (richburdon) [reply packages/core/protocols/src/proto/dxos/data.proto:85]: what's the duplication?

File: packages/core/protocols/src/proto/dxos/data.proto (hunk 1 above)
Comment (richburdon) [reply packages/core/protocols/src/proto/dxos/data.proto:29]: i'm not sure. see my issue about competing subscriptions. the worker's pipeline can only be in one state so if there are multiple subscriptions (e.g. from different clients) then hire can it handle those. similarly if the single subscription is changed hire dues the service know which of your objects to invalidate?

**Pull request closed without merging**